LibGUI+DisplaySettings: Add allowed file types to `FilePicker`

mkanilsson opened the pull request:
This allows `FileSystemModel` and `FilePicker` to take a list of allowed file extensions/types which the user can use to only show files that the program is compatible with.

This also makes DisplaySettings 'Select wallpaper' dialog make use of this feature.


https://user-images.githubusercontent.com/18418792/206281034-a59c5323-87ef-4dce-bc41-d8ceda15b842.mp4

frhun commented:
@mkanilsson Why did you close it?

mkanilsson commented:
Because I misspelled `GUI::FilePicker::ExtensionFilter` and I want to make some other changes like call the feature something along the lines of "File Types Filter" instead of "Allowed File Extensions" in the FilePicker to make it easier to understand what it actually is.

Sorry, I'm new to open source (this is my second PR) and thought it'd be easier to make the changes at my end, fixup the commits and then open a new PR, instead of cluttering this PR with new names and alike.

frhun commented:
Making use of rebase in PRs is heavily encouraged in the Serenity project!
It's fine to rewrite all your commit history, as long as it stays the same basic idea at it's core.

I would also recommend getting the classes of file filters (like image files you have here) not through static methods, but for example a dedicated namespace like `FileTypeFilters`, or something like that.

frhun commented:
Also, if all file types are allowed, I think it would be better to not show that box at all. It's just visual clutter in that case.

mkanilsson commented:
I'll do both of the things you mentioned!

Also, should I reopen the PR or make a new like I talked about?

ADKaster commented on Userland/Libraries/LibGUI/FileSystemModel.cpp:
```diff
@@ -122,8 +122,19 @@ void FileSystemModel::Node::traverse_if_needed()
             directory_children.append(move(child));
-        else
-            file_children.append(move(child));
+        else {
```
when adding braces to the else here, please add them to the corresponding `if` above to keep the entire block consistent.

mkanilsson replied:
This and the below has been fixed, should it be a new commit or amended into the older commits and then force pushed?

ADKaster replied:
Sorry, looks like I missed this comment. The fixed formatting should be fixed-up into the commit that introduced the formatting mishap.

ADKaster commented on Userland/Libraries/LibGUI/FilePicker.cpp:
```diff
@@ -113,11 +117,41 @@ FilePicker::FilePicker(Window* parent_window, Mode mode, StringView filename, St
+        file_types_filters_combo->remove_from_parent();
+    }
+
+    auto open_parent_directory_action
```
This formatting looks borked. I assume that removing the newline here still lets clang-format be happy?

ADKaster commented on Userland/Libraries/LibGUI/FilePicker.cpp:
```diff
@@ -113,11 +117,41 @@ FilePicker::FilePicker(Window* parent_window, Mode mode, StringView filename, St
+        spacer.set_fixed_height(22);
+        file_types_filter_label->remove_from_parent();
+
+        file_types_filters_combo->parent_widget()->insert_child_before(GUI::Widget::construct(), *file_types_filters_combo);
```
I'm not sure about this pattern, but if it's been used elsewhere then no comment on the pattern itself.

One question though: Shouldn't this new child have fixed height of 22 as well to preserve the old layout before the combo box existed? Or does the filter types label take care of that?

mkanilsson replied:
That code reverts the GUI to how it looked before this PR, so it removes the "Files of Type:" label with an empty widget(reverts [this](https://github.com/SerenityOS/serenity/pull/16363/files#diff-e383b1da93f7b5cd5743ced5fafe25792fb68a7f97ee2a808ba34a9ff08e7448L32)) which requires a fixed height of 22 to look correct and it also replaces the filter combo box with an empty widget(reverts [this](https://github.com/SerenityOS/serenity/pull/16363/files#diff-e383b1da93f7b5cd5743ced5fafe25792fb68a7f97ee2a808ba34a9ff08e7448L85)), which in the original GML was a `GUI::Layout::Spacer` but I couldn't find a way to insert a spacer before or after a widget so I went with a widget and everything looks fine even when height isn't specified (see image below).

![2022-12-15_19-11](https://user-images.githubusercontent.com/18418792/207937631-6dac87f5-676b-4fe5-9195-349eddb72b6c.png)

About this pattern, I'm not sure, it's the only way I can think of, if I just remove the label and combo box the GUI will be wrong (see image below),

![2](https://user-images.githubusercontent.com/18418792/207939950-d555dc3a-44ef-4a60-ae46-a89105f5cd7e.png)

ADKaster replied:
Makes sense, thanks for the explanation!

ADKaster reviewed (commented):
This is pretty cool! I have two style comments and one real question. 

I'm not thrilled about adding extra uses of DeprecatedString to the codebase, but using new String would just result in a bunch of .release_value_but_fixme_propogate_errors() due to the usage in LibGUI, so that's fine to overlook imo.

mkanilsson commented:
I'm so sorry, I think I've made a mistake when I was fixing the issues and rebased.

How do I solve this?

I made my fixes with `git rebase -i HEAD~5` and selected fixup, then I pulled SerenityOS/serenity master, rebased my branch with `git rebase master` and then I pulled my remote branch and lastly, pushed to my branch.

ADKaster reviewed (approved):
Sorry for taking so long on this, these changes look great now! I've restarted one of the CI pipelines as a sanity check, but if that fails, you might have to rebase on master before we can merge this in.
